File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
    int rows[4] = {1,0,-1,0};
    int cols[4] = {0,-1,0,1};

    bool isValid(int r, int c, int m, int n){
        return r>=0 && r<m && c>=0 && c<n;
    }

    void dfs(int i, int j, int m, int n, vector<vector<bool>>& mark, vector<vector<char>>& board){
        mark[i][j] = true;

        for(int k=0; k<4; k++)
        if(isValid(i+rows[k], j+cols[k], m, n) && board[i+rows[k]][j+cols[k]] == 'O' && mark[i+rows[k]][j+cols[k]] == false) dfs(i+rows[k],j+cols[k],m,n,mark,board);
    }

public:
    void solve(vector<vector<char>>& board) {
        size_t m = board.size(), n = board[0].size();

        vector<vector<bool>> mark(m, vector<bool>(n, false));

        for(int i=0; i<n; i++)
        if(board[0][i] == 'O' && mark[0][i] == false) dfs(0,i,m,n,mark,board);

        for(int i=1; i<m; i++)
        if(board[i][n-1] == 'O' && mark[i][n-1] == false) dfs(i,n-1,m,n,mark,board);

        for(int i=n-2; i>=0; i--)
        if(board[m-1][i] == 'O' && mark[m-1][i] == false) dfs(m-1,i,m,n,mark,board);

        for(int i=m-2; i>=1; i--)
        if(board[i][0] == 'O' && mark[i][0] == false) dfs(i,0,m,n,mark,board);

        for(int i=0; i<m; i++)
        for(int j=0; j<n; j++)
        if(board[i][j] == 'O' && mark[i][j] == false) board[i][j] = 'X';
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows, cols = len(board), len(board[0])
        visited, directions = set(), ((1,0),(0,1),(-1,0),(0,-1))
        def dfs(row: int, col: int) -> None:
            if row < 0 or col < 0 or row == rows or col == cols or (row,col) in visited or board[row][col] != 'O' : return
            visited.add((row,col))
            board[row][col] = 'T'
            for d in directions:
                n_row, n_col = row + d[0], col + d[1]
                dfs(n_row, n_col)
        for c in range(cols):
            if (0,c) not in visited: dfs(0,c)
            if (rows-1,c) not in visited: dfs(rows-1,c)
        for r in range(rows):
            if (r, 0) not in visited: dfs(r,0)
            if (r, cols-1) not in visited: dfs(r, cols-1)
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == 'O' : board[r][c] = 'X'
                elif board[r][c] == 'T' : board[r][c] = 'O'